checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -0,0 +1,284 @@
+package org.checkerframework.checker.resourceleak;
+
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableSet;
+import com.sun.source.tree.Tree;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import org.checkerframework.checker.mustcall.qual.Owning;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.common.wholeprograminference.WholeProgramInference;
+import org.checkerframework.dataflow.cfg.ControlFlowGraph;
+import org.checkerframework.dataflow.cfg.UnderlyingAST;
+import org.checkerframework.dataflow.cfg.block.Block;
+import org.checkerframework.dataflow.cfg.block.SpecialBlock;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * This class contains Resource Leak Checker annotation inference algorithm. For now, it just
+ * contains inference logic for @Owning annotations on final owning fields. It adds @Owning
+ * annotation on a field if it finds a method that satisfies @MustCall obligation of the field along
+ * some path to the regular exit point.
+ */
+public class MustCallInferenceLogic {
+
+  /**
+   * The type factory for the Resource Leak Checker, which is used to access the Must Call Checker.
+   */
+  private final ResourceLeakAnnotatedTypeFactory typeFactory;
+
+  /** The @Owning annotation. */
+  protected final AnnotationMirror OWNING;
+
+  /** The control flow graph. */
+  private ControlFlowGraph cfg;
+
+  /** The underlying AST for the method. */
+  private UnderlyingAST.CFGMethod cfgMethod;
+
+  /**
+   * Creates a MustCallInferenceLogic. If the whole program inference is not null, the type
+   * factory's postAnalyze method would instantiate a new MustCallInferenceLogic using this
+   * constructor and then call {@link #inference()}.
+   *
+   * @param typeFactory the type factory
+   * @param cfg the ControlFlowGraph
+   */
+  MustCallInferenceLogic(ResourceLeakAnnotatedTypeFactory typeFactory, ControlFlowGraph cfg) {
+    this.typeFactory = typeFactory;
+    this.cfg = cfg;
+    this.cfgMethod = (UnderlyingAST.CFGMethod) cfg.getUnderlyingAST();
+    OWNING = AnnotationBuilder.fromClass(this.typeFactory.getElementUtils(), Owning.class);
+  }
+
+  /**
+   * It tracks called methods for fields with non-empty @MustCall obligation of type {@link
+   * FieldWithCMVs} along all paths to the regular exit point in the method body. This is the main
+   * function in the MustCallInferenceLogic.
+   */
+  void inference() {
+
+    Set<BlockWithFieldsFacts> visited = new HashSet<>();
```

Comment:
Document that the only reason we need a worklist here is to handle cases where `@MustCall` has multiple methods.  It's still not clear to me that supporting that scenario is worth it at this point.